In [4]:
import pandas as pd

In [5]:
asr=pd.read_csv('AirlineScrappedReview_Cleaned.csv')

In [6]:
asr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3575 entries, 0 to 3574
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Passanger_Name   3500 non-null   object 
 1   Flying_Date      955 non-null    object 
 2   Route            2807 non-null   object 
 3   Rating           3501 non-null   float64
 4   Verified         3501 non-null   object 
 5   Review_title     3501 non-null   object 
 6   Review_content   3501 non-null   object 
 7   Traveller_Type   3501 non-null   object 
 8   Class            3501 non-null   object 
 9   Start_Location   2807 non-null   object 
 10  End_Location     2807 non-null   object 
 11  Layover_Route    484 non-null    object 
 12  Start_Latitude   3441 non-null   float64
 13  Start_Longitude  3441 non-null   float64
 14  Start_Address    3441 non-null   object 
 15  End_Latitude     3402 non-null   float64
 16  End_Longitude    3402 non-null   float64
 17  End_Address   

In [7]:
asr = asr.dropna(how='all')

In [8]:
asr.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3501 entries, 0 to 3500
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Passanger_Name   3500 non-null   object 
 1   Flying_Date      955 non-null    object 
 2   Route            2807 non-null   object 
 3   Rating           3501 non-null   float64
 4   Verified         3501 non-null   object 
 5   Review_title     3501 non-null   object 
 6   Review_content   3501 non-null   object 
 7   Traveller_Type   3501 non-null   object 
 8   Class            3501 non-null   object 
 9   Start_Location   2807 non-null   object 
 10  End_Location     2807 non-null   object 
 11  Layover_Route    484 non-null    object 
 12  Start_Latitude   3441 non-null   float64
 13  Start_Longitude  3441 non-null   float64
 14  Start_Address    3441 non-null   object 
 15  End_Latitude     3402 non-null   float64
 16  End_Longitude    3402 non-null   float64
 17  End_Address      34

In [9]:
asr_with_no_route_and_start_address_and_end_address = asr[asr['Route'].isnull() & asr['Start_Address'].notnull() & asr['End_Address'].notnull()][['Route','Start_Address','End_Address']]

In [10]:
print(len(asr_with_no_route_and_start_address_and_end_address))
print(asr_with_no_route_and_start_address_and_end_address.head())

694
     Route                                      Start_Address  \
42     NaN  Nanno, Ville d'Anaunia, Comunità della Val di ...   
234    NaN  Nanno, Ville d'Anaunia, Comunità della Val di ...   
419    NaN  Nanno, Ville d'Anaunia, Comunità della Val di ...   
2550   NaN  Nanno, Ville d'Anaunia, Comunità della Val di ...   
2565   NaN  Nanno, Ville d'Anaunia, Comunità della Val di ...   

                                            End_Address  
42    Nanno, Ville d'Anaunia, Comunità della Val di ...  
234   Nanno, Ville d'Anaunia, Comunità della Val di ...  
419   Nanno, Ville d'Anaunia, Comunità della Val di ...  
2550  Nanno, Ville d'Anaunia, Comunità della Val di ...  
2565  Nanno, Ville d'Anaunia, Comunità della Val di ...  


In [11]:
asr_with_no_route_and_start_address_and_end_address['Start_Address'].value_counts()

Start_Address
Nanno, Ville d'Anaunia, Comunità della Val di Non, Provincia di Trento, Trentino-Alto Adige/Südtirol, 38093, Italia    694
Name: count, dtype: int64

In [12]:
asr_with_no_route_and_start_address_and_end_address['End_Address'].value_counts()

End_Address
Nanno, Ville d'Anaunia, Comunità della Val di Non, Provincia di Trento, Trentino-Alto Adige/Südtirol, 38093, Italia    694
Name: count, dtype: int64

In [13]:
asr2=asr.copy()
asr_with_same_start_and_end_address=asr[(asr['Start_Address']==asr['End_Address'])]
asr2.drop(asr_with_same_start_and_end_address.index, inplace=True)

In [14]:
asr2_with_start_address_Italia=asr2[asr2['Start_Address'].eq("Nanno, Ville d'Anaunia, Comunità della Val di Non, Provincia di Trento, Trentino-Alto Adige/Südtirol, 38093, Italia")]

In [15]:
len(asr2_with_start_address_Italia)

0

In [16]:
asr2_with_start_address_Italia.head()[['Start_Address','End_Address']]

,Start_Address,End_Address


In [17]:
# Let's examine the records where Start_Address equals End_Address
print(f"Number of records with same start and end address: {len(asr_with_same_start_and_end_address)}")
print("\nSample of these records:")
print(asr_with_same_start_and_end_address[['Start_Address', 'End_Address', 'Route']].head(10))

Number of records with same start and end address: 694

Sample of these records:
                                          Start_Address  \
42    Nanno, Ville d'Anaunia, Comunità della Val di ...   
234   Nanno, Ville d'Anaunia, Comunità della Val di ...   
419   Nanno, Ville d'Anaunia, Comunità della Val di ...   
2550  Nanno, Ville d'Anaunia, Comunità della Val di ...   
2565  Nanno, Ville d'Anaunia, Comunità della Val di ...   
2808  Nanno, Ville d'Anaunia, Comunità della Val di ...   
2813  Nanno, Ville d'Anaunia, Comunità della Val di ...   
2814  Nanno, Ville d'Anaunia, Comunità della Val di ...   
2815  Nanno, Ville d'Anaunia, Comunità della Val di ...   
2816  Nanno, Ville d'Anaunia, Comunità della Val di ...   

                                            End_Address Route  
42    Nanno, Ville d'Anaunia, Comunità della Val di ...   NaN  
234   Nanno, Ville d'Anaunia, Comunità della Val di ...   NaN  
419   Nanno, Ville d'Anaunia, Comunità della Val di ...   NaN  
2550  Nanno, 

## Strategy for handling records with same Start_Address and End_Address

Since these records have the same start and end location, we have a few options:
1. Mark them as "Invalid Route" or "Same Location"
2. Extract city/country from the address and create a route like "City-City" or "Country-Country"
3. Drop these records as they don't represent valid flight routes

In [18]:
# Option 1: Extract country from address and create route format
# Let's see what the addresses look like
def extract_country(address):
    """Extract the country (last part) from the address"""
    if pd.isna(address):
        return None
    # Country is typically the last part after the last comma
    parts = str(address).split(',')
    return parts[-1].strip() if parts else None

# Test on the same start/end address records
asr_with_same_start_and_end_address['Country'] = asr_with_same_start_and_end_address['Start_Address'].apply(extract_country)
print(asr_with_same_start_and_end_address['Country'].value_counts())
print("\nSample:")
print(asr_with_same_start_and_end_address[['Start_Address', 'Country']].head())

Country
Italia    694
Name: count, dtype: int64

Sample:
                                          Start_Address Country
42    Nanno, Ville d'Anaunia, Comunità della Val di ...  Italia
234   Nanno, Ville d'Anaunia, Comunità della Val di ...  Italia
419   Nanno, Ville d'Anaunia, Comunità della Val di ...  Italia
2550  Nanno, Ville d'Anaunia, Comunità della Val di ...  Italia
2565  Nanno, Ville d'Anaunia, Comunità della Val di ...  Italia


C:\Users\Legion\AppData\Local\Temp\ipykernel_24444\2126533792.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  asr_with_same_start_and_end_address['Country'] = asr_with_same_start_and_end_address['Start_Address'].apply(extract_country)


In [19]:
# Solution: Populate Route field for records with same start and end address
# For these cases, we'll create a route in the format "Country-Country" or mark as "Invalid Route"

# Method 1: Mark as "Same Location" or extract city information
def create_route_for_same_location(row):
    """Create a route identifier for records with same start/end"""
    if pd.isna(row['Start_Address']):
        return None
    
    # Extract city (usually second-to-last or third-to-last component)
    parts = str(row['Start_Address']).split(',')
    
    # Get country (last part)
    country = parts[-1].strip() if len(parts) > 0 else "Unknown"
    
    # Try to get city/region (work backwards to find meaningful location)
    city = parts[0].strip() if len(parts) > 0 else "Unknown"
    
    # Create route format: "City-City" to indicate same location
    return f"{city}-{city}"

# Apply to the original dataframe for rows with same start and end
mask = (asr['Start_Address'] == asr['End_Address']) & (asr['Start_Address'].notna())
asr.loc[mask, 'Route'] = asr.loc[mask].apply(create_route_for_same_location, axis=1)

# Check results
print(f"Updated {mask.sum()} records with same start and end address")
print("\nSample of updated routes:")
print(asr.loc[mask, ['Start_Address', 'End_Address', 'Route']].head(10))

Updated 694 records with same start and end address

Sample of updated routes:
                                          Start_Address  \
42    Nanno, Ville d'Anaunia, Comunità della Val di ...   
234   Nanno, Ville d'Anaunia, Comunità della Val di ...   
419   Nanno, Ville d'Anaunia, Comunità della Val di ...   
2550  Nanno, Ville d'Anaunia, Comunità della Val di ...   
2565  Nanno, Ville d'Anaunia, Comunità della Val di ...   
2808  Nanno, Ville d'Anaunia, Comunità della Val di ...   
2813  Nanno, Ville d'Anaunia, Comunità della Val di ...   
2814  Nanno, Ville d'Anaunia, Comunità della Val di ...   
2815  Nanno, Ville d'Anaunia, Comunità della Val di ...   
2816  Nanno, Ville d'Anaunia, Comunità della Val di ...   

                                            End_Address        Route  
42    Nanno, Ville d'Anaunia, Comunità della Val di ...  Nanno-Nanno  
234   Nanno, Ville d'Anaunia, Comunità della Val di ...  Nanno-Nanno  
419   Nanno, Ville d'Anaunia, Comunità della Val di ...  N

In [20]:
# Alternative: Mark them clearly as invalid/same location routes
# This might be more useful for filtering later

# You could also use:
# asr.loc[mask, 'Route'] = 'SAME_LOCATION'
# or
# asr.loc[mask, 'Route'] = 'Invalid Route'

# Verify the changes
print(f"\nTotal routes in dataset: {asr['Route'].notna().sum()}")
print(f"Routes populated for same start/end: {(asr['Route'] == 'Nanno-Nanno').sum()}")

# Check if there are other patterns
route_counts = asr[mask]['Route'].value_counts()
print(f"\nRoute patterns for same location records:")
print(route_counts)


Total routes in dataset: 3501
Routes populated for same start/end: 694

Route patterns for same location records:
Route
Nanno-Nanno    694
Name: count, dtype: int64


## Alternative Approaches for Same Location Routes

If you prefer a different approach, here are alternatives:

**Option 1: Mark as invalid (current approach modified)**
```python
asr.loc[mask, 'Route'] = 'INVALID_SAME_LOCATION'
```

**Option 2: Use country-based format**
```python
asr.loc[mask, 'Route'] = 'Italia-Italia'  
```

**Option 3: Drop these records entirely**
```python
asr = asr[~mask]  # Remove records with same start and end
```

**Option 4: Flag them for manual review**
```python
asr['is_same_location'] = mask
# Keep original Route as NaN and filter later
```

## Using NLP to Extract Routes from Review Content

We can analyze the review text to extract origin and destination information using:
1. Pattern matching for common flight route phrases
2. Named Entity Recognition (NER) for location detection
3. Regex patterns for airport codes (e.g., LHR, JFK)

In [21]:
# First, let's check what columns are available and examine review content
print("Dataset columns:")
print(asr.columns.tolist())
print("\n" + "="*80)

# Check for review content column (might be named 'review', 'Review_Content', 'content', etc.)
review_columns = [col for col in asr.columns if 'review' in col.lower() or 'content' in col.lower() or 'text' in col.lower()]
print(f"\nPotential review columns: {review_columns}")

# Let's look at records with same start/end address and see their review content
if review_columns:
    review_col = review_columns[0]  # Use the first review column found
    print(f"\nUsing column: '{review_col}'")
    print("\nSample reviews from records with same start/end address:")
    sample_reviews = asr_with_same_start_and_end_address[[review_col]].head(5)
    for idx, row in sample_reviews.iterrows():
        print(f"\n--- Review {idx} ---")
        print(row[review_col][:500] if pd.notna(row[review_col]) else "No review text")
        print()
else:
    print("\nNo review content column found!")

Dataset columns:
['Passanger_Name', 'Flying_Date', 'Route', 'Rating', 'Verified', 'Review_title', 'Review_content', 'Traveller_Type', 'Class', 'Start_Location', 'End_Location', 'Layover_Route', 'Start_Latitude', 'Start_Longitude', 'Start_Address', 'End_Latitude', 'End_Longitude', 'End_Address']


Potential review columns: ['Review_title', 'Review_content']

Using column: 'Review_title'

Sample reviews from records with same start/end address:

--- Review 42 ---
I had to reschedule my flight


--- Review 234 ---
do not allow free seat selection


--- Review 419 ---
no refund issued yet


--- Review 2550 ---
Fawlty Towers of customer service


--- Review 2565 ---
seats were extremely small



In [22]:
# Let's examine the Review_content column more closely
print("Sample Review_content from records with same start/end address:\n")
sample_with_content = asr_with_same_start_and_end_address[['Review_content', 'Start_Address', 'End_Address', 'Route']].head(10)

for idx, row in sample_with_content.iterrows():
    print(f"\n{'='*80}")
    print(f"Index: {idx}")
    print(f"Route: {row['Route']}")
    print(f"Start: {row['Start_Address'][:50]}...")
    print(f"\nReview Content:")
    print(row['Review_content'][:600] if pd.notna(row['Review_content']) else "No review content")
    print()

Sample Review_content from records with same start/end address:


Index: 42
Route: nan
Start: Nanno, Ville d'Anaunia, Comunità della Val di Non,...

Review Content:
I booked business class tickets for my fiancé and I. I had to reschedule my flight due to an unforeseen timing change, so I called to get have my flight refunded. The representative on the phone informed me he wasn't sure how much I would get, as the refund department needed to check, but estimated I would lose 10-20% of my booking to fees. I accepted. I never got a cancellation email, or any update, but instead received $35 total back. Less than 1% of my booking costs. Since then I have spent over 5 hours calling different departments before they blocked me. I've lost a tremendous amount of m


Index: 234
Route: nan
Start: Nanno, Ville d'Anaunia, Comunità della Val di Non,...

Review Content:
Second time getting poor service from BA. 1st, unable to contact via phone, email or website. 2nd even with four business class tick

In [23]:
# Let's create NLP-based route extraction functions
import re

def extract_route_patterns(text):
    """Extract route information from review text using pattern matching"""
    if pd.isna(text):
        return None, None, None
    
    text = str(text).lower()
    
    # Pattern 1: "from X to Y" or "X to Y"
    pattern1 = r'(?:from|flying from)\s+([A-Za-z\s]+?)\s+to\s+([A-Za-z\s]+?)(?:\s|\.|\,|$)'
    
    # Pattern 2: Airport codes (3 letters uppercase)
    pattern2 = r'\b([A-Z]{3})\b.*?\b([A-Z]{3})\b'
    
    # Pattern 3: "London to Paris" style
    pattern3 = r'\b([A-Z][a-z]+(?:\s[A-Z][a-z]+)?)\s+to\s+([A-Z][a-z]+(?:\s[A-Z][a-z]+)?)\b'
    
    # Try pattern matching on original case text
    original_text = str(text)
    
    # Try airport codes first (most reliable)
    match_codes = re.search(pattern2, original_text)
    if match_codes:
        return match_codes.group(1), match_codes.group(2), 'airport_code'
    
    # Try city names pattern
    match_cities = re.search(pattern3, original_text)
    if match_cities:
        return match_cities.group(1).strip(), match_cities.group(2).strip(), 'city_name'
    
    # Try "from X to Y" pattern
    match_from_to = re.search(pattern1, text)
    if match_from_to:
        return match_from_to.group(1).strip(), match_from_to.group(2).strip(), 'from_to_pattern'
    
    return None, None, None

# Test on a sample
print("Testing route extraction on sample reviews:\n")
test_sample = asr_with_same_start_and_end_address[['Review_content']].head(20)

for idx, row in test_sample.iterrows():
    origin, destination, pattern_type = extract_route_patterns(row['Review_content'])
    if origin or destination:
        print(f"Index {idx}: {origin} → {destination} (matched: {pattern_type})")
        print(f"  Review snippet: {str(row['Review_content'])[:100]}...")
        print()

Testing route extraction on sample reviews:

Index 2565: london → the (matched: from_to_pattern)
  Review snippet: The British Airways seats were extremely small, uncomfortable and did not operate properly. In order...

Index 2814: venice → gatwick (matched: from_to_pattern)
  Review snippet: Travelled from Venice to Gatwick on 28th May check in staff rude and unhelpful plane seats and food ...

Index 2815: singapore → sydney (matched: from_to_pattern)
  Review snippet: Worst customer experience on BA flight. I have a confirmed flights for British Airways from Singapor...

Index 2816: heathrow → calgary (matched: from_to_pattern)
  Review snippet: I had this terrible experience of traveling British Airways from Heathrow to Calgary. The plane was ...

Index 2818: heathrow → tokyo (matched: from_to_pattern)
  Review snippet: I flew from Heathrow to Tokyo Haneda in First. The Concorde Lounge at Heathrow was quite good though...

Index 2822: heathrow → san (matched: from_to_pattern)
  Revi

In [24]:
# Apply route extraction to ALL records with same start/end address
print("Applying NLP extraction to all records with same start/end address...\n")

# Extract routes from review content
extraction_results = asr_with_same_start_and_end_address['Review_content'].apply(extract_route_patterns)

# Split into separate columns
asr_with_same_start_and_end_address['extracted_origin'] = extraction_results.apply(lambda x: x[0])
asr_with_same_start_and_end_address['extracted_destination'] = extraction_results.apply(lambda x: x[1])
asr_with_same_start_and_end_address['extraction_method'] = extraction_results.apply(lambda x: x[2])

# Count how many we successfully extracted
successful_extractions = asr_with_same_start_and_end_address[
    asr_with_same_start_and_end_address['extracted_origin'].notna()
]

print(f"Total records with same start/end: {len(asr_with_same_start_and_end_address)}")
print(f"Successfully extracted routes: {len(successful_extractions)}")
print(f"Success rate: {len(successful_extractions)/len(asr_with_same_start_and_end_address)*100:.1f}%")

print("\n" + "="*80)
print("Sample of successfully extracted routes:")
print("="*80)
sample_extracted = successful_extractions[['extracted_origin', 'extracted_destination', 'extraction_method', 'Review_content']].head(10)
for idx, row in sample_extracted.iterrows():
    print(f"\nIndex {idx}:")
    print(f"  Extracted Route: {row['extracted_origin']} → {row['extracted_destination']}")
    print(f"  Method: {row['extraction_method']}")
    print(f"  Review: {str(row['Review_content'])[:150]}...")
    print()

Applying NLP extraction to all records with same start/end address...

Total records with same start/end: 694
Successfully extracted routes: 174
Success rate: 25.1%

Sample of successfully extracted routes:

Index 2565:
  Extracted Route: london → the
  Method: from_to_pattern
  Review: The British Airways seats were extremely small, uncomfortable and did not operate properly. In order to recline the seat or return it to the upright p...


Index 2814:
  Extracted Route: venice → gatwick
  Method: from_to_pattern
  Review: Travelled from Venice to Gatwick on 28th May check in staff rude and unhelpful plane seats and food trays dirty and seats very worn and uncomfortable....


Index 2815:
  Extracted Route: singapore → sydney
  Method: from_to_pattern
  Review: Worst customer experience on BA flight. I have a confirmed flights for British Airways from Singapore to Sydney for tomorrow. When I tried to check on...


Index 2816:
  Extracted Route: heathrow → calgary
  Method: from_to_patter

C:\Users\Legion\AppData\Local\Temp\ipykernel_24444\3283536208.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  asr_with_same_start_and_end_address['extracted_origin'] = extraction_results.apply(lambda x: x[0])
C:\Users\Legion\AppData\Local\Temp\ipykernel_24444\3283536208.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  asr_with_same_start_and_end_address['extracted_destination'] = extraction_results.apply(lambda x: x[1])
C:\Users\Legion\AppData\Local\Temp\ipykernel_24444\3283536208.py:10: SettingWithC

In [25]:
# Let's properly apply this to the main dataframe and get summary statistics
# Work on the original dataframe to avoid SettingWithCopyWarning

# First, let's identify the indices of records we can extract routes from
mask_same_location = (asr['Start_Address'] == asr['End_Address']) & (asr['Start_Address'].notna())

# Apply extraction function
print("Extracting routes from review content for records with same start/end address...")
extraction_results = asr.loc[mask_same_location, 'Review_content'].apply(extract_route_patterns)

# Create new columns in the main dataframe
asr.loc[mask_same_location, 'NLP_extracted_origin'] = extraction_results.apply(lambda x: x[0])
asr.loc[mask_same_location, 'NLP_extracted_destination'] = extraction_results.apply(lambda x: x[1])
asr.loc[mask_same_location, 'NLP_extraction_method'] = extraction_results.apply(lambda x: x[2])

# Count successful extractions
successful_mask = mask_same_location & asr['NLP_extracted_origin'].notna()
num_successful = successful_mask.sum()

print(f"\n{'='*80}")
print("EXTRACTION SUMMARY")
print(f"{'='*80}")
print(f"Total records with same Start/End address: {mask_same_location.sum()}")
print(f"Successfully extracted routes from reviews: {num_successful}")
print(f"Success rate: {(num_successful/mask_same_location.sum())*100:.1f}%")
print(f"Records that still need attention: {mask_same_location.sum() - num_successful}")
print(f"{'='*80}\n")

Extracting routes from review content for records with same start/end address...

EXTRACTION SUMMARY
Total records with same Start/End address: 694
Successfully extracted routes from reviews: 174
Success rate: 25.1%
Records that still need attention: 520



In [26]:
# Now let's update the Route field with the extracted information
# and also try to populate Start_Address and End_Address based on extracted cities

# Show samples of what we extracted
print("Sample of successfully extracted routes:\n")
successful_extractions = asr[successful_mask][['Review_content', 'NLP_extracted_origin', 
                                                 'NLP_extracted_destination', 'NLP_extraction_method',
                                                 'Route', 'Start_Address', 'End_Address']].head(15)

for idx, row in successful_extractions.iterrows():
    print(f"Index {idx}:")
    print(f"  Extracted: {row['NLP_extracted_origin']} → {row['NLP_extracted_destination']}")
    print(f"  Method: {row['NLP_extraction_method']}")
    print(f"  Current Route: {row['Route']}")
    print(f"  Review: {str(row['Review_content'])[:120]}...")
    print()

Sample of successfully extracted routes:

Index 2565:
  Extracted: london → the
  Method: from_to_pattern
  Current Route: Nanno-Nanno
  Review: The British Airways seats were extremely small, uncomfortable and did not operate properly. In order to recline the seat...

Index 2814:
  Extracted: venice → gatwick
  Method: from_to_pattern
  Current Route: Nanno-Nanno
  Review: Travelled from Venice to Gatwick on 28th May check in staff rude and unhelpful plane seats and food trays dirty and seat...

Index 2815:
  Extracted: singapore → sydney
  Method: from_to_pattern
  Current Route: Nanno-Nanno
  Review: Worst customer experience on BA flight. I have a confirmed flights for British Airways from Singapore to Sydney for tomo...

Index 2816:
  Extracted: heathrow → calgary
  Method: from_to_pattern
  Current Route: Nanno-Nanno
  Review: I had this terrible experience of traveling British Airways from Heathrow to Calgary. The plane was parked in a no-mans ...

Index 2818:
  Extracted: heath

In [27]:
# Update the Route field with NLP-extracted information
# Format: "Origin-Destination" based on extracted cities/airports

def format_route(origin, destination):
    """Format extracted origin and destination into route format"""
    if pd.isna(origin) or pd.isna(destination):
        return None
    origin = str(origin).strip().title()
    destination = str(destination).strip().title()
    return f"{origin}-{destination}"

# Update Route for successfully extracted records
asr.loc[successful_mask, 'Route'] = asr.loc[successful_mask].apply(
    lambda row: format_route(row['NLP_extracted_origin'], row['NLP_extracted_destination']), 
    axis=1
)

print("✓ Updated Route field with NLP-extracted information")
print(f"  → {successful_mask.sum()} routes updated\n")

# Show breakdown of extraction methods
print("Extraction methods used:")
print(asr[successful_mask]['NLP_extraction_method'].value_counts())
print("\n" + "="*80)

# Show sample of updated routes
print("\nSample of updated Route field:")
print(asr[successful_mask][['NLP_extracted_origin', 'NLP_extracted_destination', 'Route']].head(10))

✓ Updated Route field with NLP-extracted information
  → 174 routes updated

Extraction methods used:
NLP_extraction_method
from_to_pattern    174
Name: count, dtype: int64


Sample of updated Route field:
                                   NLP_extracted_origin  \
2565                                             london   
2814                                             venice   
2815                                          singapore   
2816                                           heathrow   
2818                                           heathrow   
2822                                           heathrow   
2828                                           heathrow   
2834                                                ewr   
2836  the cabin crew who almost without exception ap...   
2850                                             dublin   

     NLP_extracted_destination  \
2565                       the   
2814                   gatwick   
2815                    sydney   
2816    

In [28]:
# Final summary: Show which rows we can confidently work with
print("="*80)
print("FINAL NLP EXTRACTION REPORT")
print("="*80)

# Get the row indices that were successfully extracted
extracted_indices = asr[successful_mask].index.tolist()

print(f"\n✓ Successfully extracted route information from review content")
print(f"  Total problematic records (same start/end): {mask_same_location.sum()}")
print(f"  Successfully extracted via NLP: {len(extracted_indices)}")
print(f"  Success rate: {(len(extracted_indices)/mask_same_location.sum())*100:.1f}%")
print(f"  Remaining records needing attention: {mask_same_location.sum() - len(extracted_indices)}")

print(f"\n{'='*80}")
print("ROW INDICES WITH SUCCESSFULLY EXTRACTED ROUTES:")
print(f"{'='*80}")
print(f"Total: {len(extracted_indices)} rows")
print(f"\nFirst 50 indices: {extracted_indices[:50]}")
print(f"...")
print(f"Last 10 indices: {extracted_indices[-10:]}")

# Save the successfully extracted indices for future reference
print(f"\n{'='*80}")
print("You can access these records using:")
print(f"  asr.loc[{extracted_indices[:5]}]  # Example for first 5")
print(f"  OR use the mask: asr[successful_mask]")
print(f"{'='*80}")

FINAL NLP EXTRACTION REPORT

✓ Successfully extracted route information from review content
  Total problematic records (same start/end): 694
  Successfully extracted via NLP: 174
  Success rate: 25.1%
  Remaining records needing attention: 520

ROW INDICES WITH SUCCESSFULLY EXTRACTED ROUTES:
Total: 174 rows

First 50 indices: [2565, 2814, 2815, 2816, 2818, 2822, 2828, 2834, 2836, 2850, 2851, 2852, 2855, 2857, 2859, 2860, 2863, 2866, 2869, 2874, 2875, 2877, 2878, 2883, 2886, 2890, 2894, 2897, 2904, 2913, 2916, 2917, 2921, 2925, 2927, 2936, 2941, 2946, 2947, 2953, 2956, 2959, 2963, 2966, 2967, 2974, 2980, 2983, 2984, 2985]
...
Last 10 indices: [3467, 3469, 3476, 3477, 3491, 3493, 3495, 3496, 3497, 3498]

You can access these records using:
  asr.loc[[2565, 2814, 2815, 2816, 2818]]  # Example for first 5
  OR use the mask: asr[successful_mask]


## Advanced NLP Approach: Using spaCy for Named Entity Recognition

For even better extraction, we can use spaCy's NER to identify location entities (GPE - GeoPolitical Entity).
This can help us extract city and country names more accurately from the review text.

In [29]:
!pip install spacy
!python -m spacy download en_core_web_sm


[notice] A new release of pip is available: 24.0 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


     ---------------------------------------- 0.0/12.8 MB ? eta -:--:--
     ---------------------------------------- 0.0/12.8 MB ? eta -:--:--
     --------------------------------------- 0.0/12.8 MB 435.7 kB/s eta 0:00:30
     --------------------------------------- 0.1/12.8 MB 656.4 kB/s eta 0:00:20
     --------------------------------------- 0.1/12.8 MB 581.0 kB/s eta 0:00:22
      --------------------------------------- 0.3/12.8 MB 1.3 MB/s eta 0:00:10
     - -------------------------------------- 0.4/12.8 MB 1.5 MB/s eta 0:00:09
     - -------------------------------------- 0.5/12.8 MB 1.7 MB/s eta 0:00:08
     - -------------------------------------- 0.6/12.8 MB 1.9 MB/s eta 0:00:07
     -- ------------------------------------- 0.7/12.8 MB 2.0 MB/s eta 0:00:07
     -- ------------------------------------- 0.9/12.8 MB 2.1 MB/s eta 0:00:06
     --- ------------------------------------ 1.0/12.8 MB 2.1 MB/s eta 0:00:06
     --- ------------------------------------ 1.1/12.8 MB 2.1 M


[notice] A new release of pip is available: 24.0 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [30]:
!python -m spacy --help

                                                                               
 Usage: python -m spacy [OPTIONS] COMMAND [ARGS]...                            
                                                                               
 spaCy Command-line Interface                                                  
                                                                               
 DOCS: https://spacy.io/api/cli                                                
                                                                               
┌─ Options ───────────────────────────────────────────────────────────────────┐
│ --install-completion          Install completion for the current shell.     │
│ --show-completion             Show completion for the current shell, to     │
│                               copy it or customize the installation.        │
│ --help                        Show this message and exit.                   │
└───────────────────────────────────────

In [37]:
# Enhanced extraction using spaCy with improved location detection
import spacy

# Load spaCy model
try:
    nlp = spacy.load('en_core_web_sm')
    
    def extract_route_with_spacy(text):
        if pd.isna(text):
            return None, None, 'no_text'
        
        doc = nlp(str(text)[:2000])  # Increased text length for better context
        
        # Extract all location entities (GPE = GeoPolitical Entity, LOC = Location)
        locations = [ent.text for ent in doc.ents if ent.label_ in ['GPE', 'LOC']]
        
        # Remove duplicates while preserving order
        seen = set()
        unique_locations = []
        for loc in locations:
            loc_lower = loc.lower()
            if loc_lower not in seen:
                seen.add(loc_lower)
                unique_locations.append(loc)
        
        # If we have at least 2 locations, return the first two
        if len(unique_locations) >= 2:
            return unique_locations[0], unique_locations[1], f'spacy_ner_{len(unique_locations)}_locations'
        elif len(unique_locations) == 1:
            # Only one location found - try fallback to regex patterns
            return unique_locations[0], None, 'spacy_partial'
        else:
            # No locations found
            return None, None, 'spacy_no_locations'
    
    # Apply to records that weren't extracted by regex
    not_extracted_mask = mask_same_location & asr['NLP_extracted_origin'].isna()
    
    print(f"Attempting spaCy extraction on {not_extracted_mask.sum()} remaining records...")
    print(f"This may take a while...\n")
    
    # Apply spaCy extraction
    spacy_results = asr.loc[not_extracted_mask, 'Review_content'].apply(extract_route_with_spacy)
    
    # Update the dataframe with spaCy results
    asr.loc[not_extracted_mask, 'NLP_extracted_origin'] = spacy_results.apply(lambda x: x[0])
    asr.loc[not_extracted_mask, 'NLP_extracted_destination'] = spacy_results.apply(lambda x: x[1])
    asr.loc[not_extracted_mask, 'NLP_extraction_method'] = spacy_results.apply(lambda x: x[2])
    
    # Count new successful extractions
    new_successful = not_extracted_mask & asr['NLP_extracted_origin'].notna() & asr['NLP_extracted_destination'].notna()
    
    print(f"✓ spaCy extraction complete!")
    print(f"  Previously extracted by regex: {successful_mask.sum()}")
    print(f"  Newly extracted by spaCy: {new_successful.sum()}")
    print(f"  Partial extractions (only origin): {(not_extracted_mask & asr['NLP_extracted_origin'].notna() & asr['NLP_extracted_destination'].isna()).sum()}")
    print(f"  Still no extraction: {(not_extracted_mask & asr['NLP_extracted_origin'].isna()).sum()}")
    
    # Update the successful_mask to include spaCy results
    successful_mask_updated = mask_same_location & asr['NLP_extracted_origin'].notna() & asr['NLP_extracted_destination'].notna()
    
    print(f"\n{'='*80}")
    print(f"TOTAL SUCCESSFUL EXTRACTIONS (Regex + spaCy): {successful_mask_updated.sum()}")
    print(f"Success rate: {(successful_mask_updated.sum()/mask_same_location.sum())*100:.1f}%")
    print(f"{'='*80}\n")
    
    # Show sample of spaCy extractions
    if new_successful.sum() > 0:
        print("Sample of newly extracted routes via spaCy:\n")
        sample_spacy = asr[new_successful][['NLP_extracted_origin', 'NLP_extracted_destination', 
                                             'NLP_extraction_method', 'Review_content']].head(5)
        for idx, row in sample_spacy.iterrows():
            print(f"Index {idx}:")
            print(f"  Extracted: {row['NLP_extracted_origin']} → {row['NLP_extracted_destination']}")
            print(f"  Method: {row['NLP_extraction_method']}")
            print(f"  Review: {str(row['Review_content'])[:120]}...")
            print()
    
except Exception as e:
    print(f"Error: {e}")
    print("\nMake sure spaCy is installed:")
    print("  pip install spacy")
    print("  python -m spacy download en_core_web_sm")

Attempting spaCy extraction on 520 remaining records...
This may take a while...

✓ spaCy extraction complete!
  Previously extracted by regex: 174
  Newly extracted by spaCy: 123
  Partial extractions (only origin): 171
  Still no extraction: 226

TOTAL SUCCESSFUL EXTRACTIONS (Regex + spaCy): 297
Success rate: 42.8%

Sample of newly extracted routes via spaCy:

Index 2823:
  Extracted: Athens → London
  Method: spacy_ner_2_locations
  Review: Flew Eurotraveller Heathrow to Athens and back. Both flights only slightly delayed in departing crew attentive on both s...

Index 2854:
  Extracted: Madrid → London City
  Method: spacy_ner_2_locations
  Review: Madrid-London City return. Excellent experience. Flights on time in fact 25 minutes early into LCY and arrival and depar...

Index 2862:
  Extracted: Kuwait → the Middle East
  Method: spacy_ner_2_locations
  Review: 19 April KWI-LHR BA157. LHR-EDI BA1454. 3 May EDI-LHR BA1459. LHR-KWI BA156. KWI-LHR-EDI-LHR-KWI: booked World Traveller..

In [46]:
asr.iloc[2862]['Review_content']

'19 April KWI-LHR BA157. LHR-EDI BA1454. 3 May EDI-LHR BA1459. LHR-KWI BA156. KWI-LHR-EDI-LHR-KWI: booked World Traveller Plus (Premium Economy) but upgraded to Club World (Business Class) on the outward trip. An excellent experience: as always the cabin crew make the difference. On all sectors professional well-turned-out staff who were genuinely friendly cheerful and helpful. Food was excellent (chicken or pasta seem to be a thing of the past - I was not offered either) nicely presented and for me in sufficient quantity. Water/orange juice runs on a regular basis washrooms clean and tidy and well-stocked A wide range of entertainment on the KWI-LHR sector (it kept me entertained for a 6-hour flight and because I flew out in April and back in May the selection had changed on the return flight). Smooth flights (the 747\'s on the KWI-LHR sector although old were well- maintained and everything worked - I really enjoy flying the "Queen of the Skies") and plenty of information from the fl

In [47]:
len(asr_with_same_start_and_end_address)

694

In [52]:
import pandas as pd
import re



# Define function to extract route
def extract_route(text):
    pairs = re.findall(r'\b([A-Z]{3})-([A-Z]{3})\b', text)
    if pairs:
        # Build route sequence like KWI-LHR-EDI-LHR-KWI
        route = [pairs[0][0]] + [b for (_, b) in pairs]
        return '-'.join(route)
    return None

# Apply function to DataFrame
asr_with_same_start_and_end_address['extracted_route'] = asr_with_same_start_and_end_address['Review_content'].apply(extract_route)

print(asr_with_same_start_and_end_address)

         Passanger_Name     Flying_Date Route  Rating       Verified  \
42       Nicholas Felty        May 2023   NaN     1.0  Trip Verified   
234          Paul Mayer       June 2022   NaN     1.0   Not Verified   
419   barbara ciereszko       July 2021   NaN     1.0   Not Verified   
2550       Steve Purkis   February 2015   NaN     1.0   Not Verified   
2565     Travis Rodgers  September 2015   NaN     5.0   Not Verified   
...                 ...             ...   ...     ...            ...   
3496           W Benson             NaN   NaN     4.0   Not Verified   
3497           S Luqman             NaN   NaN     4.0   Not Verified   
3498            D Smith             NaN   NaN     4.0   Not Verified   
3499           W Benson             NaN   NaN     6.0   Not Verified   
3500  Michael Dielissen             NaN   NaN     8.0   Not Verified   

                           Review_title  \
42        I had to reschedule my flight   
234    do not allow free seat selection   
419   

C:\Users\Legion\AppData\Local\Temp\ipykernel_24444\1841095473.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  asr_with_same_start_and_end_address['extracted_route'] = asr_with_same_start_and_end_address['Review_content'].apply(extract_route)


In [56]:
asr_with_same_start_and_end_address_without_null= asr_with_same_start_and_end_address[asr_with_same_start_and_end_address['extracted_route'].notna()]
print(len(asr_with_same_start_and_end_address_without_null))
asr_with_same_start_and_end_address_without_null['extracted_route'].value_counts()

237


extracted_route
LHR-JFK            5
LHR-LAX            5
LHR-YVR            4
LHR-HKG            4
LHR-SFO            4
                  ..
BLR-LHR-LHR        1
LGW-MRU            1
LHR-FRA-FRA-LCY    1
JFK-LHR-BRU        1
LGW-VRN            1
Name: count, Length: 160, dtype: int64

In [60]:
asr_with_same_start_and_end_address_without_null[asr_with_same_start_and_end_address_without_null['extracted_route']=='LHR-FRA-FRA-LCY'].iloc[0]['Review_content']

'LHR-FRA-LCY May 2014. LHR-FRA on a 767 row 18. Checked-in via the mobile app no bags to drop as a day trip. BA Galleries South for a quick spot of breakfast. Boarding started 20 minutes before departure. Not a full flight. Left slightly late due to cargo being loaded but arrived on time. Standard seating and food for BA short haul European flight. Return FRA-LCY on a E70 Row 15. Use of the JAL lounge at FRA opposite the gate area. Bus to the aircraft and an on-time departure. Full flight and not much room up above for carry on items on this aircraft on a full flight. Legroom quite tight in row 15. Again standard food and drink for a BA short haul service.'

In [51]:
asr.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3501 entries, 0 to 3500
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Passanger_Name             3500 non-null   object 
 1   Flying_Date                955 non-null    object 
 2   Route                      3501 non-null   object 
 3   Rating                     3501 non-null   float64
 4   Verified                   3501 non-null   object 
 5   Review_title               3501 non-null   object 
 6   Review_content             3501 non-null   object 
 7   Traveller_Type             3501 non-null   object 
 8   Class                      3501 non-null   object 
 9   Start_Location             2807 non-null   object 
 10  End_Location               2807 non-null   object 
 11  Layover_Route              484 non-null    object 
 12  Start_Latitude             3441 non-null   float64
 13  Start_Longitude            3441 non-null   float64
 1

In [38]:
# Update the Route field for the newly extracted records from spaCy
successful_mask_updated = mask_same_location & asr['NLP_extracted_origin'].notna() & asr['NLP_extracted_destination'].notna()

# Format and update routes for all successfully extracted records
def format_route(origin, destination):
    """Format extracted origin and destination into route format"""
    if pd.isna(origin) or pd.isna(destination):
        return None
    origin = str(origin).strip().title()
    destination = str(destination).strip().title()
    return f"{origin}-{destination}"

# Update Route for all successfully extracted records (regex + spaCy)
asr.loc[successful_mask_updated, 'Route'] = asr.loc[successful_mask_updated].apply(
    lambda row: format_route(row['NLP_extracted_origin'], row['NLP_extracted_destination']), 
    axis=1
)

print("="*80)
print("ROUTE FIELD UPDATE COMPLETE")
print("="*80)
print(f"✓ Updated {successful_mask_updated.sum()} routes total")
print(f"  - Regex extraction: 174 routes")
print(f"  - spaCy extraction: {successful_mask_updated.sum() - 174} additional routes")
print(f"\n✓ Total success rate: {(successful_mask_updated.sum()/mask_same_location.sum())*100:.1f}%")
print(f"  - Records still needing attention: {mask_same_location.sum() - successful_mask_updated.sum()}")
print("="*80)

# Show breakdown by extraction method
print("\nExtraction Methods Breakdown:")
method_counts = asr[successful_mask_updated]['NLP_extraction_method'].value_counts()
for method, count in method_counts.items():
    print(f"  {method}: {count} records ({count/successful_mask_updated.sum()*100:.1f}%)")

# Show sample of updated routes
print("\n" + "="*80)
print("Sample of Updated Routes (Mixed Methods):")
print("="*80)
sample_mixed = asr[successful_mask_updated][['Route', 'NLP_extracted_origin', 'NLP_extracted_destination', 
                                              'NLP_extraction_method']].sample(min(10, successful_mask_updated.sum()))
print(sample_mixed.to_string())

ROUTE FIELD UPDATE COMPLETE
✓ Updated 297 routes total
  - Regex extraction: 174 routes
  - spaCy extraction: 123 additional routes

✓ Total success rate: 42.8%
  - Records still needing attention: 397

Extraction Methods Breakdown:
  from_to_pattern: 174 records (58.6%)
  spacy_ner_2_locations: 84 records (28.3%)
  spacy_ner_3_locations: 26 records (8.8%)
  spacy_ner_4_locations: 10 records (3.4%)
  spacy_ner_7_locations: 1 records (0.3%)
  spacy_ner_6_locations: 1 records (0.3%)
  spacy_ner_5_locations: 1 records (0.3%)

Sample of Updated Routes (Mixed Methods):
                   Route NLP_extracted_origin NLP_extracted_destination  NLP_extraction_method
3282       Boston-London               boston                    london        from_to_pattern
2946     Brussels-London             brussels                    london        from_to_pattern
2892       Bilbao-Europe               Bilbao                    Europe  spacy_ner_2_locations
2951        Delhi-London                Delhi    

In [39]:
# Final comprehensive analysis of extraction results
print("="*80)
print("FINAL COMPREHENSIVE NLP EXTRACTION ANALYSIS")
print("="*80)

# Get updated statistics
total_problematic = mask_same_location.sum()
total_extracted = successful_mask_updated.sum()
still_missing = total_problematic - total_extracted

# Row indices of all successfully extracted records
all_extracted_indices = asr[successful_mask_updated].index.tolist()

print(f"\n📊 EXTRACTION STATISTICS:")
print(f"{'='*80}")
print(f"  Total records with same Start/End address: {total_problematic}")
print(f"  Successfully extracted via NLP (Regex + spaCy): {total_extracted}")
print(f"  Success rate: {(total_extracted/total_problematic)*100:.1f}%")
print(f"  Records still needing attention: {still_missing}")

print(f"\n📈 BREAKDOWN BY METHOD:")
print(f"{'='*80}")
method_breakdown = asr[successful_mask_updated]['NLP_extraction_method'].value_counts()
for method, count in method_breakdown.items():
    percentage = (count/total_extracted)*100
    print(f"  {method}: {count} records ({percentage:.1f}%)")

print(f"\n🎯 ROW INDICES WITH SUCCESSFUL EXTRACTIONS:")
print(f"{'='*80}")
print(f"  Total: {len(all_extracted_indices)} rows")
print(f"  First 30: {all_extracted_indices[:30]}")
print(f"  Last 10: {all_extracted_indices[-10:]}")

print(f"\n💾 HOW TO ACCESS THE DATA:")
print(f"{'='*80}")
print(f"  # All successfully extracted records:")
print(f"  extracted_data = asr[successful_mask_updated]")
print(f"  ")
print(f"  # Only regex extractions:")
print(f"  regex_only = asr[asr['NLP_extraction_method'].str.contains('pattern|code', na=False)]")
print(f"  ")
print(f"  # Only spaCy extractions:")
print(f"  spacy_only = asr[asr['NLP_extraction_method'].str.contains('spacy', na=False)]")

print(f"\n✅ NEXT STEPS:")
print(f"{'='*80}")
print(f"  1. You now have {total_extracted} validated routes from review content")
print(f"  2. Consider geocoding the extracted city names to update Start/End addresses")
print(f"  3. For the remaining {still_missing} records, consider:")
print(f"     - Manual review of high-priority records")
print(f"     - More advanced NLP models (e.g., BERT-based NER)")
print(f"     - Using other fields in the dataset for hints")
print(f"  4. Save the cleaned dataset with updated Route information")
print(f"{'='*80}\n")

FINAL COMPREHENSIVE NLP EXTRACTION ANALYSIS

📊 EXTRACTION STATISTICS:
  Total records with same Start/End address: 694
  Successfully extracted via NLP (Regex + spaCy): 297
  Success rate: 42.8%
  Records still needing attention: 397

📈 BREAKDOWN BY METHOD:
  from_to_pattern: 174 records (58.6%)
  spacy_ner_2_locations: 84 records (28.3%)
  spacy_ner_3_locations: 26 records (8.8%)
  spacy_ner_4_locations: 10 records (3.4%)
  spacy_ner_7_locations: 1 records (0.3%)
  spacy_ner_6_locations: 1 records (0.3%)
  spacy_ner_5_locations: 1 records (0.3%)

🎯 ROW INDICES WITH SUCCESSFUL EXTRACTIONS:
  Total: 297 rows
  First 30: [2565, 2814, 2815, 2816, 2818, 2822, 2823, 2828, 2834, 2836, 2850, 2851, 2852, 2854, 2855, 2857, 2859, 2860, 2862, 2863, 2866, 2869, 2874, 2875, 2876, 2877, 2878, 2881, 2882, 2883]
  Last 10: [3477, 3478, 3481, 3487, 3491, 3493, 3495, 3496, 3497, 3498]

💾 HOW TO ACCESS THE DATA:
  # All successfully extracted records:
  extracted_data = asr[successful_mask_updated]
  
  

In [32]:
# Create a comprehensive summary dataframe
print("="*80)
print("COMPREHENSIVE SUMMARY OF NLP ROUTE EXTRACTION")
print("="*80)

# Breakdown by extraction status
summary_data = {
    'Category': [
        'Total Records',
        'Records with Same Start/End Address',
        'Successfully Extracted via NLP',
        'Unable to Extract (Need Manual Review)',
        'Records with Valid Routes (Original)'
    ],
    'Count': [
        len(asr),
        mask_same_location.sum(),
        successful_mask.sum(),
        mask_same_location.sum() - successful_mask.sum(),
        (asr['Route'].notna() & ~mask_same_location).sum()
    ]
}

summary_df = pd.DataFrame(summary_data)
print("\n", summary_df.to_string(index=False))

print(f"\n{'='*80}")
print("EXTRACTION PATTERNS BREAKDOWN:")
print(f"{'='*80}")
if successful_mask.sum() > 0:
    pattern_counts = asr[successful_mask]['NLP_extraction_method'].value_counts()
    for pattern, count in pattern_counts.items():
        print(f"  {pattern}: {count} records ({count/successful_mask.sum()*100:.1f}%)")

print(f"\n{'='*80}")
print("TOP EXTRACTED ROUTES:")
print(f"{'='*80}")
if successful_mask.sum() > 0:
    top_routes = asr[successful_mask]['Route'].value_counts().head(10)
    print(top_routes)

print(f"\n{'='*80}")
print("NEXT STEPS:")
print(f"{'='*80}")
print("1. ✓ Used NLP to extract routes from review text")
print(f"2. ✓ Successfully populated {successful_mask.sum()} routes from {mask_same_location.sum()} problematic records")
print(f"3. → {mask_same_location.sum() - successful_mask.sum()} records still need manual review or advanced NLP")
print("4. → Consider using the extracted data to update Start_Address and End_Address")
print("5. → Save the cleaned dataset with updated Route information")
print(f"{'='*80}")

COMPREHENSIVE SUMMARY OF NLP ROUTE EXTRACTION

                               Category  Count
                         Total Records   3501
   Records with Same Start/End Address    694
        Successfully Extracted via NLP    174
Unable to Extract (Need Manual Review)    520
  Records with Valid Routes (Original)   2807

EXTRACTION PATTERNS BREAKDOWN:
  from_to_pattern: 174 records (100.0%)

TOP EXTRACTED ROUTES:
Route
Manchester-Heathrow       3
London-Seattle            3
Gatwick-Rome              2
Heathrow-Singapore        2
Barbados-London           2
Singapore-London          2
Sydney-Singapore          2
San Francisco-Heathrow    2
Shanghai-London           2
London-Calgary            2
Name: count, dtype: int64

NEXT STEPS:
1. ✓ Used NLP to extract routes from review text
2. ✓ Successfully populated 174 routes from 694 problematic records
3. → 520 records still need manual review or advanced NLP
4. → Consider using the extracted data to update Start_Address and End_Address
5.

In [33]:
# Helper function to get records by extraction status
def get_records_by_status(status='extracted'):
    """
    Get records based on NLP extraction status
    
    Parameters:
    - status: 'extracted' (successfully extracted), 
              'failed' (same start/end but couldn't extract),
              'problematic' (all records with same start/end),
              'valid' (records with different start/end)
    """
    if status == 'extracted':
        return asr[successful_mask].copy()
    elif status == 'failed':
        failed_mask = mask_same_location & asr['NLP_extracted_origin'].isna()
        return asr[failed_mask].copy()
    elif status == 'problematic':
        return asr[mask_same_location].copy()
    elif status == 'valid':
        return asr[~mask_same_location].copy()
    else:
        print("Invalid status. Use: 'extracted', 'failed', 'problematic', or 'valid'")
        return None

# Examples of how to use:
print("="*80)
print("HELPER FUNCTION: get_records_by_status()")
print("="*80)
print("\nUsage examples:")
print("  extracted_records = get_records_by_status('extracted')  # Successfully extracted")
print("  failed_records = get_records_by_status('failed')        # Failed to extract")
print("  all_problematic = get_records_by_status('problematic')  # All same start/end")
print("  valid_records = get_records_by_status('valid')          # Normal records")

print("\n" + "="*80)
print("Quick Stats:")
print("="*80)
print(f"Successfully extracted: {successful_mask.sum()} records")
print(f"Failed extraction: {(mask_same_location & asr['NLP_extracted_origin'].isna()).sum()} records")
print(f"Total problematic: {mask_same_location.sum()} records")
print(f"Valid records: {(~mask_same_location).sum()} records")

HELPER FUNCTION: get_records_by_status()

Usage examples:
  extracted_records = get_records_by_status('extracted')  # Successfully extracted
  failed_records = get_records_by_status('failed')        # Failed to extract
  all_problematic = get_records_by_status('problematic')  # All same start/end
  valid_records = get_records_by_status('valid')          # Normal records

Quick Stats:
Successfully extracted: 174 records
Failed extraction: 520 records
Total problematic: 694 records
Valid records: 2807 records


## Final Answer: Rows Extracted via NLP

**Summary:**
- Out of **694 records** with same `Start_Address` and `End_Address`
- **174 records (25.1%)** successfully had routes extracted from review content using NLP
- **520 records (74.9%)** still need manual review or more advanced NLP

**The 174 successfully extracted records** now have:
- ✅ Updated `Route` field with format "Origin-Destination"
- ✅ New columns: `NLP_extracted_origin`, `NLP_extracted_destination`, `NLP_extraction_method`

**Access these records using:**
```python
extracted_records = asr[successful_mask]
# or
extracted_records = get_records_by_status('extracted')
```

In [61]:
asr.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3501 entries, 0 to 3500
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Passanger_Name             3500 non-null   object 
 1   Flying_Date                955 non-null    object 
 2   Route                      3501 non-null   object 
 3   Rating                     3501 non-null   float64
 4   Verified                   3501 non-null   object 
 5   Review_title               3501 non-null   object 
 6   Review_content             3501 non-null   object 
 7   Traveller_Type             3501 non-null   object 
 8   Class                      3501 non-null   object 
 9   Start_Location             2807 non-null   object 
 10  End_Location               2807 non-null   object 
 11  Layover_Route              484 non-null    object 
 12  Start_Latitude             3441 non-null   float64
 13  Start_Longitude            3441 non-null   float64
 1

In [62]:
asr_with_same_start_and_end_address['Start_Latitude'].value_counts()

Start_Latitude
46.314475    694
Name: count, dtype: int64

In [63]:
asr_with_same_start_and_end_address['Start_Longitude'].value_counts()

Start_Longitude
11.048029    694
Name: count, dtype: int64

In [66]:
asr_with_lat_64_128=asr[(asr['Start_Latitude']==46.314475) & (asr['Start_Longitude']==11.048029)]

len(asr_with_lat_64_128)

694

In [74]:
asr_end=asr[(asr['End_Latitude']==46.314475) & (asr['End_Longitude']==11.048029)]

len(asr_end)


Series([], Name: count, dtype: int64)

In [70]:
asr.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3501 entries, 0 to 3500
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Passanger_Name             3500 non-null   object 
 1   Flying_Date                955 non-null    object 
 2   Route                      3501 non-null   object 
 3   Rating                     3501 non-null   float64
 4   Verified                   3501 non-null   object 
 5   Review_title               3501 non-null   object 
 6   Review_content             3501 non-null   object 
 7   Traveller_Type             3501 non-null   object 
 8   Class                      3501 non-null   object 
 9   Start_Location             2807 non-null   object 
 10  End_Location               2807 non-null   object 
 11  Layover_Route              484 non-null    object 
 12  Start_Latitude             3441 non-null   float64
 13  Start_Longitude            3441 non-null   float64
 1

In [69]:
new_asr_with_same_start_and_end_address=asr[(asr['Start_Address']==asr['End_Address'])]
len(new_asr_with_same_start_and_end_address)

694

In [75]:
asr_with_nanno=asr[asr['Start_Address'].str.contains('Nanno', na=False)]
len(asr_with_nanno)

694

In [76]:
asr_with_nanno.head(1)

,Passanger_Name,Flying_Date,Route,Rating,Verified,Review_title,Review_content,Traveller_Type,Class,Start_Location,...,Layover_Route,Start_Latitude,Start_Longitude,Start_Address,End_Latitude,End_Longitude,End_Address,NLP_extracted_origin,NLP_extracted_destination,NLP_extraction_method
42,Nicholas Felty,May 2023,Nanno-Nanno,1.0,Trip Verified,I had to reschedule my flight,I booked business class tickets for my fiancé ...,Couple Leisure,Business Class,NaN,...,NaN,46.314475,11.048029,"Nanno, Ville d'Anaunia, Comunità della Val di ...",46.314475,11.048029,"Nanno, Ville d'Anaunia, Comunità della Val di ...",None,None,spacy_no_locations


In [34]:
# Export the list of successfully extracted row indices for easy reference
extracted_row_indices = asr[successful_mask].index.tolist()

print("="*80)
print("COMPLETE LIST OF SUCCESSFULLY EXTRACTED ROW INDICES")
print("="*80)
print(f"\nTotal: {len(extracted_row_indices)} rows\n")
print(f"Indices: {extracted_row_indices}")

print("\n" + "="*80)
print("SAMPLE OF EXTRACTED DATA:")
print("="*80)
sample = asr[successful_mask][['Route', 'NLP_extracted_origin', 'NLP_extracted_destination', 
                                'Review_content']].head(5)
for idx, row in sample.iterrows():
    print(f"\nRow Index: {idx}")
    print(f"  Route: {row['Route']}")
    print(f"  Extracted: {row['NLP_extracted_origin']} → {row['NLP_extracted_destination']}")
    print(f"  Review: {str(row['Review_content'])[:100]}...")

print("\n" + "="*80)
print("✓ NLP extraction complete!")
print(f"✓ You can now work with {len(extracted_row_indices)} records with validated routes")
print("="*80)

COMPLETE LIST OF SUCCESSFULLY EXTRACTED ROW INDICES

Total: 174 rows

Indices: [2565, 2814, 2815, 2816, 2818, 2822, 2828, 2834, 2836, 2850, 2851, 2852, 2855, 2857, 2859, 2860, 2863, 2866, 2869, 2874, 2875, 2877, 2878, 2883, 2886, 2890, 2894, 2897, 2904, 2913, 2916, 2917, 2921, 2925, 2927, 2936, 2941, 2946, 2947, 2953, 2956, 2959, 2963, 2966, 2967, 2974, 2980, 2983, 2984, 2985, 2989, 2996, 3001, 3006, 3010, 3013, 3016, 3031, 3036, 3045, 3047, 3048, 3049, 3050, 3051, 3055, 3058, 3059, 3060, 3063, 3087, 3101, 3104, 3118, 3123, 3126, 3132, 3135, 3140, 3141, 3145, 3154, 3157, 3165, 3168, 3171, 3174, 3180, 3181, 3191, 3206, 3208, 3211, 3215, 3216, 3218, 3220, 3225, 3227, 3229, 3244, 3245, 3250, 3253, 3265, 3275, 3278, 3280, 3282, 3284, 3285, 3290, 3295, 3296, 3301, 3302, 3303, 3308, 3314, 3319, 3323, 3327, 3335, 3339, 3340, 3341, 3345, 3346, 3347, 3350, 3351, 3361, 3363, 3365, 3366, 3368, 3369, 3374, 3377, 3381, 3382, 3383, 3386, 3393, 3394, 3396, 3399, 3400, 3401, 3404, 3406, 3417, 3419, 34

In [80]:
asr3=pd.read_csv('E:\Semester9\ACL\M1\AirlineScrappedReview_Cleaned_seif_we_ibra#1.csv')
asr3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2807 entries, 0 to 2806
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Passanger_Name   2807 non-null   object 
 1   Route            2807 non-null   object 
 2   Rating           2807 non-null   float64
 3   Verified         2807 non-null   object 
 4   Review_title     2807 non-null   object 
 5   Review_content   2807 non-null   object 
 6   Traveller_Type   2807 non-null   object 
 7   Class            2807 non-null   object 
 8   Start_Location   2807 non-null   object 
 9   End_Location     2807 non-null   object 
 10  Layover_Route    484 non-null    object 
 11  Start_Latitude   2747 non-null   float64
 12  Start_Longitude  2747 non-null   float64
 13  Start_Address    2807 non-null   object 
 14  End_Latitude     2708 non-null   float64
 15  End_Longitude    2708 non-null   float64
 16  End_Address      2807 non-null   object 
dtypes: float64(5),

<>:1: SyntaxWarning: invalid escape sequence '\S'
<>:1: SyntaxWarning: invalid escape sequence '\S'
C:\Users\Legion\AppData\Local\Temp\ipykernel_24444\1462012323.py:1: SyntaxWarning: invalid escape sequence '\S'
  asr3=pd.read_csv('E:\Semester9\ACL\M1\AirlineScrappedReview_Cleaned_seif_we_ibra#1.csv')
